UploadDaemon/Upload/IUpload.cs

```diff
@@ -36,5 +36,14 @@ public interface IUpload
         /// Must not contain passwords/access keys/...
         /// </summary>
         string Describe();
+
+        /// <summary>
+        /// Returns an object that implements GetHashCode and Equals which can
+        /// be used as a key in a Dictionary that identifies this upload. I.e.
+        /// two keys from the same type of IUpload should only be equal if
+        /// they represent the same upload destination and coverage targeted
+        /// at both IUploads can be merged and uploaded in one go.
+        /// </summary>
+        object GetDictionaryKey();
```

Comment:
I don't like the technical name. Would `TargetId` or `DestinationId` describe the concept? I think if I would see we cluster by `upload.TargetId`, I would understand that we cluster by where the coverage goes to. Then we could shorten the comment to say that two `IUpload`s with the same target should return the same `TargetId`. I think the comment should not state what we use this for, because this might change. In fact, equal `TargetId`s are not sufficient for merging, since also the `Revision` must be the same. This key really only models the upload-target concept.

Reply:
I agree with the rename but disagree with removing that part of the comment. Someone implementing a new type of upload needs to understand the implications of this method's implementation. "Target" is not specific enough to make someone understand what the implications for the line coverage merge process are. E.g. I think the intuitive implementation for the TS upload would be to just return all TeamcsaleServer fields, because that's the target right? But the username and access key should not factor into the line coverage merge decision, so that would not be correct. But that's not transparent to someone who doesn't know the merge logic (or even that it exists).
I also disagree that comments need to be generic enough so you'll never have to change them in the future, i.e. take every possible future use case into account. Comments - like any part of the source code - have to be maintained by us which includes adjusting them if they are no longer accurate. Just like identifiers have to be renamed when they no longer accurately reflect their underlying concepts.
Let's discuss in person if you are not convinced!

Reply:
If you have doubts that somebody else would infer the right consequences from the name alone, I'm fine with leaving the additional information. Better to err on the side of too much details, I guess.

I absolutely agree that comments need to be maintained. Unfortunately, in my experience, this is more easily forgotten than updating code, because an outdated comment doesn't (directly) break any functionality. Therefore, I try to place the comments close to where I would need to do code changes that would require to update the comment, such that I'm likely to see them when I change the code and they are likely to appear in the review diff. Therefore, I'm concerned about a getter that documents how the return value is used. It's easy to change the usage (or add new, different usages) without every seeing/updating the comment. I would prefer to document the properties of the return value, independent of the usage.

How about something like: "Two `IUpload` should return equal target Ids if, and only if, the result of uploading coverage through either of them will be the same, i.e., the coverage will be uploaded to the same destination and it will be indistinguishable which of the uploads was used, after the fact."

Reply:
Sorry for writing this out instead of calling you, it was easier to formulate a draft in typing 😇

Reply:
Sounds good, I like it